File: kotlin-typescript/karakum/nameResolvers/resolveFunctionReturnTypeItemName.js
import ts from "typescript";
import * as karakum from "karakum";

export default (node, context) => {
    const typeScriptService = context.lookupService(karakum.typeScriptServiceKey)
    const getParent = typeScriptService?.getParent.bind(typeScriptService) ?? (node => node.parent)

    let arrayType = getParent(node)
    if (ts.isParenthesizedTypeNode(arrayType)) {
        arrayType = getParent(arrayType)
    }
    if (!arrayType) return null
    if (!ts.isArrayTypeNode(arrayType)) return null

    let functionNode = getParent(arrayType)
    if (ts.isTypeOperatorNode(functionNode)) {
        functionNode = getParent(functionNode)
    }
    if (!functionNode) return null
    if (!ts.isFunctionDeclaration(functionNode)) return null
    if (functionNode.name === undefined) return null
    if (
        functionNode.type !== arrayType
        && functionNode.type !== arrayType.parent
    ) return null

    const parentName = functionNode.name.text

    return `${karakum.capitalize(parentName)}ResultItem`
}

File: kotlin-typescript/karakum/src/jsMain/kotlin/typescript/karakum/nameResolvers/resolveFunctionReturnTypeItemName.kt
package typescript.karakum.nameResolvers

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import typescript.Node
import typescript.isArrayTypeNode
import typescript.isFunctionDeclaration
import typescript.isParenthesizedTypeNode
import typescript.isTypeOperatorNode

fun resolveFunctionReturnTypeItemName(node: Node, context: Context) = nullable {
    val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

    val arrayType = ensureNotNull(
        typeScriptService.getParent(node)
            ?.let {
                if (isParenthesizedTypeNode(it)) {
                    typeScriptService.getParent(it)
                } else it
            }
    )
    ensure(isArrayTypeNode(arrayType))

    val function = ensureNotNull(
        typeScriptService.getParent(arrayType)
            ?.let {
                if (isTypeOperatorNode(it)) {
                    typeScriptService.getParent(it)
                } else it
            }
    )
    ensure(isFunctionDeclaration(function))
    ensure(
        function.type == arrayType
                || function.type == arrayType.parent
    )
    val parentName = ensureNotNull(function.name).text

    "${parentName.replaceFirstChar { it.titlecase() }}ResultItem"
}
